### PR TITLE
Full Site Editing: Introduce a template editing mode inside the post editor

### PR DESCRIPTION
**Pull request by youknowriad**
This  PR explores the addition of a template editing mode right inside the "post editor". Allowing users to zoom-out to the template level.

**Todo**

 - [x] Template resolution for new posts and pages (auto-drafts) tries to resolve the **404** template instead of the **single** one. I believe the root issue is that we rely on frontend resolution from Core which renders 404 for these posts. Not sure exactly how to solve that yet. I created a dedicate issue for this #27391 
 - [x] I noticed that sometimes (maybe because of auto-saves), the post gets saved with the content of the template inside post_content.
 - [x] Post Content block should be usable inside the post editor in template mode which mean, I remove the check we had for recursions. We might have to find a different way to guard about the inclusion of such blocks inside the post content.
 - [x] Add an e2e test.

### CODE REVIEW

File: packages/edit-post/src/components/header/template-zoom-toggle/index.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+	return (
+		<Button
+			isPressed={ isActive }
+			onClick={ () => {
+				toggleFeature( 'templateZoomOut' );
+			} }
+			icon={ stack }
+			/* translators: button label text should, if possible, be under 16
+characters. */
```
Comment (mcsf) [inline packages/edit-post/src/components/header/template-zoom-toggle/index.js:25]: We aren't complying with this. :P

File: packages/edit-post/src/components/visual-editor/index.js [hunk 2]
```diff
@@ -43,9 +49,11 @@ function VisualEditor() {
 				<CopyHandler>
 					<WritingFlow>
 						<ObserveTyping>
-							<div className="edit-post-visual-editor__post-title-wrapper">
-								<PostTitle />
-							</div>
+							{ ! templateZoomOut && (
```
Comment (mcsf) [inline packages/edit-post/src/components/visual-editor/index.js:52]: For now, this is fine if we just want to see how the feature feels, but in the long run we should absorb this exception. Off the top of my head, we might altogether remove `PostTitle` from here and ensure that, within `BlockList`, there is a `PostTitle` at the head. I don't know if this is an acceptable "breaking" change in terms of DOM and classes — though, IMO, these have always been private interfaces.

File: packages/edit-post/src/utils/find-template.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+				resolved: true,
+				slug: data.post_name,
+			} )
+		 )[ 0 ];
+	}
+
+	return newTemplate;
+}
+
+findTemplate.siteUrl = '';
```
Comment (mcsf) [inline packages/edit-post/src/utils/find-template.js:39]: What's this for?

File: packages/edit-post/src/utils/find-template.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * Find the template for a given page path.
+ *
+ * @param {string}   path The page path.
+ * @param {Function} getEntityRecords The promise-returning `getEntityRecords` selector to use.
+ *
+ * @return {number} The found template ID.
+ */
+export default async function findTemplate( path, getEntityRecords ) {
```
Comment (mcsf) [inline packages/edit-post/src/utils/find-template.js:19]: We aren't passing `getEntityRecords` in `editor.js`, fwiw

File: packages/block-library/src/post-content/edit.js [hunk 5]
```diff
@@ -13,19 +12,11 @@ import PostContentInnerBlocks from './inner-blocks';
 export default function PostContentEdit( {
 	context: { postId: contextPostId, postType: contextPostType },
 } ) {
-	const { id: currentPostId, type: currentPostType } = useSelect(
-		( select ) => select( 'core/editor' ).getCurrentPost() ?? {}
-	);
 	const blockProps = useBlockProps();
 
 	// Only render InnerBlocks if the context is different from the active post
 	// to avoid infinite recursion of post content.
```
Comment (mcsf) [inline packages/block-library/src/post-content/edit.js:18]: Comment needs update

File: packages/block-library/src/post-title/edit.js [hunk 6]
```diff
@@ -33,31 +34,35 @@ export default function PostTitleEdit( {
...
-				postId
-			),
-		[ postType, postId ]
+	const [ title, setTitle ] = useEntityProp(
+		'postType',
+		postType,
+		'title',
+		postId
 	);
+	const [ link ] = useEntityProp( 'postType', postType, 'title', postId );
```
Comment (mcsf) [inline packages/block-library/src/post-title/edit.js:43]: > `'title'`

Is this right?

File: packages/block-library/src/post-title/edit.js [hunk 7]
```diff
@@ -33,31 +34,35 @@ export default function PostTitleEdit( {
...
-	if ( ! post ) {
-		return null;
-	}
+	let titleElement = (
+		<PlainText
+			tagName={ TagName }
+			placeholder={ __( 'No Title' ) }
+			{ ...blockProps }
+			value={ title }
+			onChange={ setTitle }
```
Comment (mcsf) [inline packages/block-library/src/post-title/edit.js:57]: It's nice to see entity props shining. :)

File: packages/edit-post/src/editor.js [hunk 8]
```diff
@@ -28,31 +27,82 @@ import preventEventDiscovery from './prevent-event-discovery';
...
 
-class Editor extends Component {
-	constructor() {
-		super( ...arguments );
+function useTemplate( link ) {
+	const getEntityRecords = useSelect(
+		( select ) => select( 'core' ).getEntityRecords,
+		[]
+	);
+	const [ template, setTemplate ] = useState( null, getEntityRecords );
```
Comment (mcsf) [inline packages/edit-post/src/editor.js:37]: What is this second argument to `useState`?

**Review by mcsf — COMMENTED** [review]
Nice one!

In terms of feature this is already interesting. It also highlights some of the current problems with FSE, like how long it takes for certain template blocks to load.

In terms of implementation, there are things we can revise, but I know this is still just a draft. :)

File: packages/edit-post/src/components/visual-editor/index.js (hunk 2 above)
Comment (youknowriad) [reply packages/edit-post/src/components/visual-editor/index.js:52]: I tried it, there's another range of issues to deal with when we do that :P

File: packages/edit-post/src/utils/find-template.js (hunk 3 above)
Comment (youknowriad) [reply packages/edit-post/src/utils/find-template.js:39]: This file is copied as is from edit-site and I believe it's very bad code :P, this should move to `core-data` (selector + resolver)

**Comment by youknowriad** [comment]
@noahtallen @Addison-Stavlo and others. The last commit here c233dbf is I believe a good refactoring for the findTemplate adhoc function that is used right now in edit-site package. It might deserve to extract it from this PR to see how possible it is to use that new selector for edit-site as well.

**Comment by youknowriad** [comment]
I also noticed that trying to resolve a template for an "auto-draft" post or page results in the 404 template being returned. I'm not sure this is expected to me, It seems like a bug in how the template resolution work?

**Comment by jasmussen** [comment]
This is a cool PR! This could be how most people will edit templates, or create new ones. 

As for how this might work, I need to think about this some more, but because of the work that needs to be done to make this happen, I wanted to take a stab at mocking up my first instinct, which is that editing templates in this way could work a little bit like how "master slides" (the equivalent of templates) are edited in Keynote:

keynote

To translate that to the block editor: 

- You're editing a post
- You press a button, "Edit [type] template", perhaps in the document sidebar
- When pressed, you enter a template editing mode, which needs to be confirmed with a "Done" button

Following that instinct, here's my first draft. First, click to edit the template:

Block Editor

Now you're editing the template, and you have to press "Done" before you are allowed to publish:

Block Editor-1

I'm not feeling all of the ingredients, and I would think this could use some iteration. Notably I'm not sure it's clear enough that you switched "modes". However I think that the _flow_ is right, and for the purposes of testing this out further in the PR, I think this might work as a V1.

One additional question, and perhaps also relevant in a larger FSE context: should the post content become locked / un-editable in this mode? 

Let me know what you think.

**Comment by youknowriad** [comment]
@jasmussen I agree that this is a good starting point and you raise some good questions. Let's try some of it and see how it feels.

**Comment by youknowriad** [comment]
@jasmussen I implemented a first iteration of that flow, I'm certain that it's buggy but I'd appreciate some feedback :)

**Comment by jasmussen** [comment]
Alright, this is cool. This is what I see:

flow

As you can see, I can't actually exit that mode without making a change. But other than that, I feel like this is _generally_ a good and relatively basic way to explore this feature. 👍 👍

Small nitpicks:

Screenshot 2020-11-26 at 10 43 47

It bugs me the link isn't the same blue as my "Modern" theme color :D I know that's not part of this PR, but it always stands out to me.

Screenshot 2020-11-26 at 10 47 46

Showing this snackbar was a good instinct!

I think we can reduce the text a bit. How about:

> Editing template. Changes made here affect all posts and pages that include the template.

**Comment by youknowriad** [comment]
I think we may need a "cancel" link next to "done" instead of allowing to return from the save when there are no changes.

**Comment by jasmussen** [comment]
> I think we may need a "cancel" link next to "done" instead of allowing to return from the save when there are no changes.

In that case I'd change the "Done" label to say "Apply".

**Comment by gziolo** [comment]
This is cool, great team work @youknowriad and @jasmussen ❤️

File: packages/block-library/src/post-content/edit.js [hunk 9]
```diff
@@ -13,19 +12,11 @@ import PostContentInnerBlocks from './inner-blocks';
 export default function PostContentEdit( {
 	context: { postId: contextPostId, postType: contextPostType },
 } ) {
-	const { id: currentPostId, type: currentPostType } = useSelect(
-		( select ) => select( 'core/editor' ).getCurrentPost() ?? {}
-	);
```
Comment (youknowriad) [inline packages/block-library/src/post-content/edit.js:18]: This reverts #24010 

It seems the checks were not good enough IMO. It relies on something that doesn't give us semantic information about the context but it was more an implementation detail.

**Comment by jameskoster** [comment]
> One additional question, and perhaps also relevant in a larger FSE context: should the post content become locked / un-editable in this mode?

I think content should be locked whilst editing a template, and the template locked whilst editing content. In light of this, there are several "states" that blocks like post content will need to entertain. I opened https://github.com/WordPress/gutenberg/issues/27064 to discuss this further.

Moving between content / template editing should be simple enough in the UI, but I think they need to _feel_ like separate exercises in order to ground the overall FSE experience. The above will help with that. 

For now it's a bit confusing how in this PR, when I switch to editing the template, I still see all the document information in the sidebar. This flow is going to be one of the most critical elements to get right in the broader FSE context. I think what we have in the PR is ok to start, but as you said there will likely be lots of refinement to do from here.

File: packages/block-library/src/post-content/edit.js (hunk 9 above)
Comment (noahtallen) [reply packages/block-library/src/post-content/edit.js:18]: Yeah, I agree that these original checks were not great, but there simply wasn't another way to do it with the information available at the time. There didn't seem to be a super clear way to add the necessary information to the context either... The problem is really, given a global context with a postID, how do you prevent nested context consumers from rendering recursively infinitely?

**Comment by youknowriad** [comment]
It is becoming clear that the editor package store and EditorProvider component are not well architectured to handle an editor that switches the edited entity. I believe we need a refactor there. (I leave this separate)

**Comment by youknowriad** [comment]
I think I fixed most of the bugs that will allow us to land a first iteration of this.
There are a number of follow-ups to be performed here:

 - Refactor the editor package store and provider.
 - Refactor the EntitiesSavedState component and pre-saving checks.
 - Solve the auto-draft template resolution issue #27391

What do you think is missing @jasmussen to land a v1 of this in terms of flows?

File: packages/core-data/src/controls.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+export function regularFetch( url ) {
```
Comment (mtias) [inline packages/core-data/src/controls.js:1]: what's regular fetch :D ?

Comment (youknowriad) [reply packages/core-data/src/controls.js:1]: don't look :P 

I tracked this here #27391 basically we don't have a REST API endpoint to resolve a template, we use a hack (by fetching the actual frontend page with an extra flag)

Comment (gziolo) [reply packages/core-data/src/controls.js:1]: 🙈 🙉 😃

File: packages/base-styles/_z-index.scss [hunk 11]
```diff
@@ -71,9 +71,6 @@ $z-layers: (
 	".edit-site-sidebar": 100000,
 	".edit-widgets-sidebar": 100000,
 	".edit-post-layout .edit-post-post-publish-panel": 100001,
-	// For larger views, the wp-admin navbar dropdown should be at top of
```
Comment (gziolo) [inline packages/base-styles/_z-index.scss:74]: What was the reason for removing this style? It seems unrelated.

**Comment by jasmussen** [comment]
> What do you think is missing @jasmussen to land a v1 of this in terms of flows?

This is what I see:

this

Because the button to edit the template is inside the sidebar, it is visually very low-impact. The template part being edited that shows up in the center of the top toolbar shares DNA with what's in the site editor (and honestly could be the same thing), and it serves as a mode change indication. Making edits and then pressing "Done" also feels like a good initial flow.

As far as next steps, I feel like the only hiccup is when I go to edit a template and _don't_ make a change, then I can't really. I think we need that cancel button you were suggesting. Someting like this:
Screenshot 2020-12-01 at 12 17 40

I don't totally love that — there's something about the verbiage and the fact that he buttons are both blue that feels off. But I think it might be worth trying, and that we'll be smarter for it. What do you think?

File: packages/e2e-tests/config/setup-debug-artifacts.js [hunk 12]
```diff
@@ -1,54 +0,0 @@
-/**
```
Comment (gziolo) [inline packages/e2e-tests/config/setup-debug-artifacts.js:1]: Why was it removed?

**Comment by youknowriad** [comment]
@jasmussen I added a "cancel" button with different styling, let me know what you think.

**Comment by jasmussen** [comment]
Screenshot 2020-12-01 at 12 43 07

I actually have a small preference for two blue buttons compared to this. But if you make the "cancel" button have blue text, like other tertiary buttons, fine to ship ;)

File: packages/e2e-tests/config/setup-debug-artifacts.js (hunk 12 above)
Comment (youknowriad) [reply packages/e2e-tests/config/setup-debug-artifacts.js:1]: haha, this makes it harder to know the exact reasons of a failing test locally, I'll restore it.

File: packages/base-styles/_z-index.scss (hunk 11 above)
Comment (youknowriad) [reply packages/base-styles/_z-index.scss:74]: i believe this z-index is not right when you have multiple sidebars, but I'll restore it for now.

File: packages/block-library/src/post-content/edit.js (hunk 9 above)
Comment (youknowriad) [reply packages/block-library/src/post-content/edit.js:18]: It seems we need a context value to indicate the post content id we're currently rendering (and not just the "post id").

**Review by gziolo — COMMENTED** [review]
The changes proposed in this PR make sense on the code level. I'm not that familiar with APIs specific to FSE so this part should be confirmed by someone else.

I really like how you were able to reuse `BlockList` component to render the template by applying changes inside `EditorProvider`. I see one challenge here moving forward – do we want to have the same set of blocks loaded when editing the post content vs the template. I would say that this is something that needs to be decided quite soon. An immediate example would be the Classic block which we should prepare for deprecation, but there are other blocks that make sense when editing the template but should never be used inside the post content like the Post Content or Template Part (that can contain Post Content) block being the most obvious ones.

File: packages/core-data/src/resolvers.js [hunk 13]
```diff
@@ -382,3 +386,38 @@ export function* getAutosaves( postType, postId ) {
 export function* getAutosave( postType, postId ) {
 	yield controls.resolveSelect( 'core', 'getAutosaves', postType, postId );
 }
+
+/**
+ * Retrieve the frontend template used for a given link.
+ *
+ * @param {string} link  Link.
+ */
+export function* __experimentalGetTemplateForLink( link ) {
```
[the remark this answers is not in the document]
Comment (youknowriad) [reply packages/core-data/src/resolvers.js:395]: That's a frontend link (whatever that link maybe, it can be a post link, a home page link, an archive page link...) you give  it a link, it gives you the template used for that link.

File: packages/e2e-tests/specs/experiments/post-editor-template-mode.test.js [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+	saveDraft,
+	trashAllPosts,
+	openPreviewPage,
+	openDocumentSettingsSidebar,
+} from '@wordpress/e2e-test-utils';
+
+describe( 'Post Editor Template mode', () => {
+	beforeAll( async () => {
+		await activateTheme( 'twentytwentyone-blocks' );
+		await trashAllPosts( 'wp_template' );
```
[the remark this answers is not in the document]
Comment (youknowriad) [reply packages/e2e-tests/specs/experiments/post-editor-template-mode.test.js:17]: Not really, a lot of these tests (including this one) create templates and save them which can impact other tests.

**Comment by gziolo** [comment]
Have you tried loading the template page from the Templates page:

Screen Shot 2020-12-01 at 14 43 07

I'm using Q theme and I can't edit the template that got updated through the edit post page:

Screen Shot 2020-12-01 at 14 47 22

File: packages/edit-post/src/components/header/index.js [hunk 15]
```diff
@@ -92,7 +100,7 @@ function Header( { setEntitiesSavedStatesCallback } ) {
 				) }
 				{ showIconLabels && ! isLargeViewport && (
 					<MoreMenu showIconLabels={ showIconLabels } />
-				) }
+				) }{ ' ' }
```
[the remark this answers is not in the document]
Comment (youknowriad) [reply packages/edit-post/src/components/header/index.js:103]: that's prettier, I didn't add it myself, I'll remove it.

**Comment by youknowriad** [comment]
>  I see one challenge here moving forward – do we want to have the same set of blocks loaded when editing the post content vs the template

This also applies to the site editor, blocks should probably have some kind of "context" where they can be used in (inside post content or not, in the widgets screen or not...)

File: packages/e2e-tests/specs/experiments/post-editor-template-mode.test.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (gziolo) [reply packages/e2e-tests/specs/experiments/post-editor-template-mode.test.js:17]: That's a good point. I didn't think about it. It's getting more tricky to run those FSE tests 😄

**Comment by youknowriad** [comment]
> So it might be not related to this PR, but it's something with how templates are handled in general.
> Edit: it looks like it's related to changes in this PR. I tried loading single template with the master branch and it works without any issues.

So the "draft" thing shouldn't happen. I had this early in the PR but now it solved. That said, the freezing when loading the template in the templates UI is also happening to me when editing the template in the site editor, it doesn't seem specific to this PR. I believe we should remove that listing entirely at some point, it was just an early thing where we didn't have any UI to access templates.

File: packages/edit-post/src/components/header/index.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (mtias) [reply packages/edit-post/src/components/header/index.js]: sure it wasn't uglify?

**Comment by jameskoster** [comment]
@jasmussen One small thought... I'm wondering if we could get away with a single button rather than the Apply/Cancel. Looking back at the Keynote example, couldn't we just have a single "Done" button? It would never be disabled, and behave differently based on whether there are changes or not:

* No changes = returns you directly to the content editor
* Changes = Save flow, then returns you to the content editor

**Comment by jasmussen** [comment]
> I'm wondering if we could get away with a single button rather than the Apply/Cancel.

Yes I think that'd be ideal. This seems fine as a starting point, it lets us validate some of the ideas and test it out. For round 2, I'd love to see a single done button, and perhaps the locking of post content when editing templates, and maybe even bringing in this template tool from the site editor, in place of the centered text label currently there:

Screenshot 2020-12-02 at 08 42 37

File: packages/edit-post/src/components/header/template-save-button/index.js [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+					.__experimentalGetTemplateInfo,
+				getEditedEntityRecord: select( coreStore )
+					.getEditedEntityRecord,
+			};
+		},
+		[]
+	);
+	return (
+		<>
+			<Button onClick={ () => setIsEditingTemplate( false ) } isTertiary>
```
Comment (gziolo) [inline packages/edit-post/src/components/header/template-save-button/index.js:37]: Should it else undo all unsaved changes applied to the template?

**Review by gziolo — COMMENTED** [review]
> Changes = Save flow, then returns you to the content editor

How about when I play with the template and I don't like changes applied, how do I discard them when there is no `Cancel` button?

File: packages/edit-post/src/components/header/template-save-button/index.js (hunk 16 above)
Comment (youknowriad) [reply packages/edit-post/src/components/header/template-save-button/index.js:37]: I'm not sure, I think it's something we should try and adapt.

File: packages/block-library/src/post-content/edit.js (hunk 9 above)
Comment (youknowriad) [reply packages/block-library/src/post-content/edit.js:18]: It's not fixed entirely, but the important breakage in the template editor is fixed by 46724a8

**Comment by jasmussen** [comment]
> How about when I play with the template and I don't like changes applied, how do I discard them when there is no Cancel button?

To me, the problem is that "review changes" has been collapsed into almost nothingness:

Screenshot 2020-12-02 at 10 39 25

I feel like this list should be expanded by default, redesigned if need be, but be a key part of the experience of editing templates, not something we bury:

Screenshot 2020-12-02 at 10 40 03

I don't see the point in this being hidden at all, to be honest. And when visible, you just uncheck your changes when you press Done.

**Comment by jameskoster** [comment]
Agreed @jasmussen. 

Another option is to simply undo until you're back to the original state, which raises another question (probably beyond this PR) – is the undo action aware of this state change, as it is in Keynote? If not, should it be?

**Comment by jasmussen** [comment]
> Another option is to simply undo until you're back to the original state, which raises another question (probably beyond this PR) – is the undo action aware of this state change, as it is in Keynote? If not, should it be?

Yep, undo is a very important ingredient. 

One thing I will add though, which is why it's so important to show that list of changes _by default, not collapsed_, is that you may accidentally make a change without knowing you did, so you don't know to undo. By showing it in that expanded list, it makes you aware of it.

**Comment by youknowriad** [comment]
I'm going to open a separate PR for the "EntititiesSavedState" component and make it expanded by default.

**Comment by jasmussen** [comment]
Awesome. Followup question: does it need to be able to collapse at all? IMO: no.

**Comment by youknowriad** [comment]
@jasmussen I  don't know, we can try whatever you like :P

**Comment by jameskoster** [comment]
@youknowriad would it be possible to add the site editing context (or at least the css classes) when you switch to template editing? That way we'll inherit the changes in https://github.com/WordPress/gutenberg/pull/27271 which make interacting with template parts more transparent.

**Review by gziolo — APPROVED** [review]
I think code-wise we are good to go. There is still a ton of work ahead, but we need to split it into smaller chunks to make it possible to iterate quicker.

**Comment by youknowriad** [comment]
@jameskoster I didn't know about that change, looking at it now, and as implemented right now, it is an ad-hoc CSS. I believe this is better addressed with an explicit setting on the block-editor package. I shall explore that separately.
